library-aux/src/main/scala/scala/runtime/Statics.scala

```diff
@@ -34,56 +34,35 @@ object Statics {
   }
 
   def longHash(lv: Long): Int = {
-    lv.toInt
-    /*
-    val iv = lv.toInt | 0
+    val iv = lv.toInt
     if (iv == lv) iv
-    else (lv ^ (lv >>> 32)).toInt | 0
-    */
+    else iv ^ (lv >>> 32).toInt
   }
 
   def doubleHash(dv: Double): Int = {
-    dv.toInt
-    /*
-    val iv = dv.toInt | 0
-    if (iv == dv)
-      return iv
-
-    val fv = dv.toFloat
-    if (fv == dv)
-      return java.lang.Float.floatToIntBits(fv)
-
-    val lv = dv.toLong
-    if (lv == dv)
-      return lv.toInt | 0
-
-    val lv2 == java.lang.Double.doubleToLongBits(dv)
-    return (lv2 ^ (lv2 >>> 32)).toInt | 0
-    */
+    val iv = dv.toInt
+    if (iv == dv) {
+      iv
+    } else {
+      // don't test the case dv.toFloat == dv
+      val lv = dv.toLong
+      if (lv == dv)
+        lv.hashCode()
```

Comment:
Oh no. I see now that the JVM impl does `lv.toInt` in that case. I actually believe that's a bug in the JVM version ^^ Because it wouldn't return something consistent with `longHash`.

I'll have to try and observe the bug tomorrow.